### DISCUSSION TITLE
Check if a function has side effects or out of scope references?

### DISCUSSION THREAD

**Discussion by zhangyx1998** [babel/babel#16204] [Q&A]
I am trying to create a utility that is capable of passing JS functions through IPC channels ([rpc-magic-proxy](https://www.npmjs.com/package/rpc-magic-proxy)). It is currently capable of creating a proxy back to the process that 'owns' the function.

However, that approach will make all proxied functions async, since IPC communications are asynchronous by nature. Therefore, I want to push it further by serializing "simple" functions and reconstructing them on another side.

For example:

```js
// Can be safely serialized
function add(x, y) { return x + y }

// Can NOT be serialized, because `z` is out of scope.
let z;
function sub(x, y) { z = x - y }
```

**Comment by j4k0xb** [comment]
https://astexplorer.net/#/gist/7283141e13dab314521744603a95e9b7/778177f43609eda9dc75df629a49649025fac118
but imo it's easier to do something like https://playwright.dev/docs/evaluating#evaluation-argument than checking for all sorts of side effects and references
